gelf/tcpreader.go

```diff
@@ -13,43 +14,90 @@ type TCPReader struct {
 	messages chan []byte
 }
 
-func newTCPReader(addr string) (*TCPReader, chan string, error) {
+type conn_channels struct {
+	drop    chan string
+	confirm chan string
+}
+
+func newTCPReader(addr string) (*TCPReader, chan string, chan string, error) {
 	var err error
 	tcpAddr, err := net.ResolveTCPAddr("tcp", addr)
 	if err != nil {
-		return nil, nil, fmt.Errorf("ResolveTCPAddr('%s'): %s", addr, err)
+		return nil, nil, nil, fmt.Errorf("ResolveTCPAddr('%s'): %s", addr, err)
 	}
 
 	listener, err := net.ListenTCP("tcp", tcpAddr)
 	if err != nil {
-		return nil, nil, fmt.Errorf("ListenTCP: %s", err)
+		return nil, nil, nil, fmt.Errorf("ListenTCP: %s", err)
 	}
 
 	r := &TCPReader{
 		listener: listener,
 		messages: make(chan []byte, 100), // Make a buffered channel with at most 100 messages
 	}
 
-	signal := make(chan string, 1)
+	close_signal := make(chan string, 1)
+	done_signal := make(chan string, 1)
 
-	go r.listenUntilCloseSignal(signal)
+	go r.listenUntilCloseSignal(close_signal, done_signal)
 
-	return r, signal, nil
+	return r, close_signal, done_signal, nil
 }
 
-func (r *TCPReader) listenUntilCloseSignal(signal chan string) {
-	defer func() { signal <- "done" }()
-	defer r.listener.Close()
+func (r *TCPReader) accepter(connections chan net.Conn) {
 	for {
 		conn, err := r.listener.Accept()
 		if err != nil {
 			break
 		}
-		go handleConnection(conn, r.messages)
+		connections <- conn
+	}
+}
+
+func (r *TCPReader) listenUntilCloseSignal(close_signal chan string, done_signal chan string) {
+	defer func() { done_signal <- "done" }()
+	defer r.listener.Close()
+	var conns []conn_channels
+	connections_channel := make(chan net.Conn, 1)
+	go r.accepter(connections_channel)
+	for {
+		select {
+		case conn := <-connections_channel:
+			drop_signal := make(chan string, 1)
+			drop_confirm := make(chan string, 1)
+			channels := conn_channels{drop: drop_signal, confirm: drop_confirm}
+			go handleConnection(conn, r.messages, drop_signal, drop_confirm)
+			conns = append(conns, channels)
+		default:
+		}
+
 		select {
-		case sig := <-signal:
+		case sig := <-close_signal:
 			if sig == "stop" {
-				break
+				if len(conns) >= 1 {
```

Comment:
Those 2 blocks being similar, could you refactor in a way which doesn't duplicate code?